Dev/bap wps onboarding

tomereli opened the pull request:
Add support bAPs for allowing other agents to onboard the prplMesh agent using wireless backhaul:
- Use Intel hostapd mesh mode instead of multi_ap for VAPs configuration
- Add WPS PBC implementation using hostapd upstream support
- Remove backhaul vaps unused code

PR in prplwrt iwlwav - https://git.prpl.dev/prplmesh/iwlwav/-/merge_requests/12

Commit:
bwl: dwpal: use intel mesh_mode

Intel hostapd doesn't support upstream hostapd multi_ap configuration
due to a hardware bug which does not allow configuring VLANs.
Instead, Intel provides "mesh_mode" configuration which allows
configuring either fAP or bAP (default is bAP).
In bAP mode, only 4addr stations which include the WFA multiap vendor
element (0x1B) will be accepted.

Move multiap configuration to a static function which first sets
start_disabled=0, then handles each supported bss_type.

In addition, add set_multiap_wps() private function which sets WPS
onboarding according to the easymesh specification if a backhaul and
fronthaul VAPs are configured. If only fronthaul VAP is configured, only
WPS for fronthaul will be supported.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ucc_listener: add stubs for start_wps_registration

Preparative commit.
Add stubs for start_wps_registration command in the controller and
agent.
In the controller, this will not be implemented since it is not required
by the test plan.
The agent implementation will be done in the followup commits.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: add cACTION_BACKHAUL_START_WPS_PBC_REQUEST

Preparative commit.
Add cACTION_BACKHAUL_START_WPS_PBC_REQUEST and
ACTION_BACKHAUL_START_WPS_PBC_REQUEST to backhaul manager yamls which
will be used to trigger WPS PBC on the 2.4G or 5G radios.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: update autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
btlvf: add cACTION_APMANAGER_START_WPS_PBC_REQUEST

Preparative commit.
Add cACTION_APMANAGER_START_WPS_PBC_REQUEST and
ACTION_APMANAGER_START_WPS_PBC_REQUEST which will be used to trigger WPC
PBS on a radio.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: add backhaul start_wps_pbc and datapath

Preparative commit.
Add start_wps_pbc API in the backhaul manager which is called by the
agent ucc listener on startWpsPbc command, and pass the message till it
gets to the ap manager.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -169,6 +169,27 @@ bool agent_ucc_listener::send_cmdu_to_destination(ieee1905_1::CmduMessageTx &cmd
+    if (method->second != "PBC") {
+        err_string = "invalid WpsConfigMethod, supporting only PBC";
+        return false;
+    }
```
(nitpick) I think it's better to do this in beerocks_ucc_listener. We do that for the other commands as well.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -131,3 +131,6 @@ cACTION_BACKHAUL_CLIENT_DISCONNECTED_NOTIFICATION:
   bssid: sMacAddr
+
+cACTION_BACKHAUL_START_WPS_PBC_REQUEST:
+  _type: class
```
Newline at end of file.

More importantly though: the band will definitely need to be part of this message. Or perhaps the radio_uid. Or the bssid. For sure, *something* :-)

tomereli replied:
Fixed the newline.
Add radio id and bssid on which to trigger wps.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -169,6 +169,27 @@ bool agent_ucc_listener::send_cmdu_to_destination(ieee1905_1::CmduMessageTx &cmd
+    if (band == params.end()) {
+        err_string = "missing band parameter";
+        return false;
+    }
```
In my version of the test plan, there is an `interface` parameter, not a `band` parameter...

But... This is f**ing marvelous... UCC indeed uses the band parameter, not the interface parameter. So the test plan is apparently just informational :-(

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -3438,4 +3438,30 @@ bool backhaul_manager::get_neighbor_links(
     return true;
 }
 
+bool backhaul_manager::start_wps_pbc(bool is_5ghz)
```
I would like to avoid (new) boolean values for the band, because at some point we might support 60GHz, and I'm not sure if 6GHz is going to be treated as a separate band or not.

So, prefer to use an enum or maybe even bitmask here.

That said, I think this `band` is just an artefact of UCC. I think in the high-level API, we'll want to be able to start PBC either on all radios or on a single radio, or maybe a list of radios. So I think it's better that we resolve the radio(s) in the ucc listener and that backhaul_manager::start_wps_pbc takes the radio(s) as argument.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -186,8 +186,7 @@ bool agent_ucc_listener::handle_start_wps_registration(
     }
-    // TODO trigger WPS PBC on the selected band
-    return true;
+    return m_backhaul_manager_ctx.start_wps_pbc(band->second == "5G");
```
If it returns false, err_string should be set to something.

arnout commented on common/beerocks/bwl/include/bwl/ap_wlan_hal.h:
```diff
@@ -346,6 +346,15 @@ class ap_wlan_hal : public virtual base_wlan_hal {
+     * 
+     * @return true on success or false on error
+     */
+    virtual bool start_wps_pbc() = 0;
```
In general, it is possible (and likely) that WPS is started on several radios simultaneously. In that case, we need to make sure that they use the same UUID. I think the only way to do that is to generate the UUID from prplmesh and pass it here.

Note that we don't need to implement generating the UUID yet - that would be an M3 feature. However, for the bwl it is better to have the interface as complete as possible already from the start - bwl changes are annoying for integrators. So, please allow a UUID as an optional parameter here.

tomereli replied:
I don't understand - the hostapd API doesn't support adding UUID, so where will it be used? Also what is the UUID type? I think this should be done in a separate PR anyway.

arnout commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -342,6 +342,12 @@ bool ap_wlan_hal_dummy::get_vap_enable(const std::string &iface_name, bool &enab
 
+bool ap_wlan_hal_dummy::start_wps_pbc()
+{
+    // TODO - simulate client connected?
```
For dummy, the minimal implementation is to write something to the log. Then you don't need a todo.

arnout reviewed (changes requested):
Looks good to me so far. I think actually this is enough for a PR to get merged right away, even if the dwpal implementation is incomplete or doesn't work yet.

That said, I would like to re-review the final version, so changes requested.

Commit:
ap manager: move backhaul vap list and enum to bwl

Preparative commit.
Move enum eBackhaulVapType and struct backhaul_vap_list_element_t to
base_wlan_hal_types.h from ap manager thread, so we can update the
backhaul VAPs when updating VAP credentials.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bcl: add FREQ_60G enum value

Preparative commit.
Add 60G frequency to the eFreqType enum, currently unused, but should at
least be defined.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: remove backhaul vaps dead code

Delete unused code related to backhaul vaps configuration and
enable/disable. This code assumes the backhaul vaps configuration is
stored in the platform DB, which is incorrect for EasyMesh since the
configuration should be received from the controller, and each VAP can
support both bAP and fAP configuration.

Since previously each bAP could support only 1 IRE, the agent logic was
to enable a new bAP every time an IRE connects to the current bAP, so
effectively there is always one active bAP beaconing.

This is again not needed, since the driver (which will be integrated
soon enough) can support any number of stations, so remove this code as
well.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: bwl: add wps pbc impl and datapath

Add start_wps_pbc() API to bwl which starts wps on the given interface,
and update the agent to call the newly added API when the request
is received in the ap manager datapath.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ucc listener: refactor wps registration

Refactor so that parsing the parameters is done in the base class.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
Re-opened since with this, autoconfiguration seems to work on the backhaul VAP, compared to https://github.com/prplfoundation/prplMesh/pull/1168 which fails. Will rebase this PR to see if the reason is something with the recent prplMesh changes.

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -3592,4 +3592,45 @@ bool backhaul_manager::add_ap_he_capabilities(const sRadioInfo &radio_info)
+        return std::string();
+    }
+
+    auto slave = *it;
```
Why do you need another variable in the way?

vitalybu commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -47,6 +47,22 @@ class backhaul_manager : public btl::transport_socket_thread {
+     * @param freq radio frequency to search
+     * @return radio mac of the found slave if found, otherwise empty string
+     */
+    const std::string freq_to_radio_mac(enum eFreqType freq) const;
```
```suggestion
    const std::string freq_to_radio_mac(eFreqType freq) const;
```

vitalybu commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -946,10 +875,34 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+            });
+        if (it == vaps.end()) {
+            LOG(ERROR) << "Failed to find VAP with bssid " << request->bssid();
+            return false;
```
I don't think you should return `false` here (it's for critical errors only).
If there's no response message, just `break`.

vitalybu commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -946,10 +875,34 @@ bool ap_manager_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_
+        auto vap_name = utils::get_iface_string_from_iface_vap_ids(
+            ap_wlan_hal->get_radio_info().iface_name, vap_id);
+
+        ap_wlan_hal->start_wps_pbc(vap_name);
```
This call can also fail

vitalybu commented on common/beerocks/bwl/econet/ap_wlan_hal_econet.h:
```diff
@@ -43,8 +43,9 @@ class ap_wlan_hal_dummy : public base_wlan_hal_dummy, public ap_wlan_hal {
-    virtual bool update_vap_credentials(
-        std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list) override;
+    virtual bool
+    update_vap_credentials(std::list<son::wireless_utils::sBssInfoConf> &bss_info_conf_list,
```
Forgot the `start_wps_pbc` override:
```c++
virtual bool start_wps_pbc(const std::string &iface_name) override;
```

vitalybu commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,8 +934,99 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+                                 WSC::eWscVendorExtSubelementBssType bss_type,
+                                 std::list<bwl::backhaul_vap_list_element_t> &backhaul_vaps_list)
+{
+    std::string bssid, ifname;
```
(nitpick)
```suggestion
    std::string bssid;
```

vitalybu commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -934,8 +934,99 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+        LOG(ERROR) << "Failed to get VAP BSSID";
+        return false;
+    }
+    if (!hostapd_config_get_value(vap_hostapd_config, "bss", ifname)) {
```
(nitpick)
```suggestion

    std::string ifname;
    if (!hostapd_config_get_value(vap_hostapd_config, "bss", ifname)) {
```

tomereli commented:
Obsolete - replaced by #1168
